src/main/java/nl/tudelft/ewi/devhub/server/backend/ConsecutiveBuildFailureBackend.java

```diff
@@ -0,0 +1,118 @@
+package nl.tudelft.ewi.devhub.server.backend;
+
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+import nl.tudelft.ewi.devhub.server.database.controllers.Commits;
+import nl.tudelft.ewi.devhub.server.database.controllers.Deliveries;
+import nl.tudelft.ewi.devhub.server.database.entities.Assignment;
+import nl.tudelft.ewi.devhub.server.database.entities.Commit;
+import nl.tudelft.ewi.devhub.server.database.entities.Delivery;
+import nl.tudelft.ewi.devhub.server.database.entities.Group;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Checks whether between the previous assignments delivery and the current deliver, or the start of time if it's the first assignment
+ * any commits through any branches that are reachable from the commit that has been handed in contained 4 or more consecutive build failures.
+ */
+public class ConsecutiveBuildFailureBackend {
+    public static final int MAX_CONSECUTIVE_BUILD_FAILURES = 4;
+    private Commits commits;
+    private Deliveries deliveries;
+
+    @Inject
+    public ConsecutiveBuildFailureBackend(Commits commits, Deliveries deliveries) {
+        this.commits = commits;
+        this.deliveries = deliveries;
+    }
+
+    public boolean hasConsecutiveBuildFailures(Group group, Delivery delivery) {
+        final List<List<Commit>> commitsToCheck = commitsToCheckForBuildFailures(group, delivery);
+        return consecutiveBuildFailures(commitsToCheck);
+    }
+
+    protected static boolean consecutiveBuildFailures(List<List<Commit>> commitsToCheck) {
+        return commitsToCheck.stream().anyMatch(ConsecutiveBuildFailureBackend::hasFourFailingCommits);
+    }
+
+
+    protected static boolean hasFourFailingCommits(List<Commit> commits) {
+        int consecutive = 0;
+
+        for (Commit commit : commits) {
+            if (commit.getBuildResult().hasFailed()) {
+                consecutive++;
+                if (consecutive == MAX_CONSECUTIVE_BUILD_FAILURES) {
```

Comment:
This traversal treats merge commits invalidly. A failing merge commit for which the first child commit actually succeeded will break this chain. We have to treat this chain as a graph during traversal and look for any failing child commit instead, which is why I would look into the option of traversing generated `SuccessiveBuildFailureWarnings`.